### PR TITLE
disambiguate two exprts with same ID

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**
The first commit is cherry-picked from #2646. I thought it deserved to be merged, even if the rest of that PR is not being worked on. I also implemented a suggestion from @smowton on that PR to make a new class `is_invalid_pointer_exprt`. The same could be done for many other pointer predicates, but I do not have time right now.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/irep_ids.def [hunk 1]
```diff
@@ -217,7 +217,7 @@ IREP_ID_ONE(invalid)
 IREP_ID_TWO(C_invalid_object, #invalid_object)
 IREP_ID_ONE(pointer_offset)
 IREP_ID_ONE(pointer_object)
-IREP_ID_TWO(is_invalid_pointer, is-invalid-pointer)
+IREP_ID_TWO(is_invalid_pointer, is_invalid_pointer)
```
Comment (smowton) [inline src/util/irep_ids.def]: IREP_ID_ONE

**Review by smowton — APPROVED** [review]
Looks good except one minor correction

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e4d3ac6d3b2f39f23ebac06dc9a206de50fd98e3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106816492

**Comment by tautschnig** [comment]
> The first commit is cherry-picked from #2646. I thought it deserved to be merged, even if the rest of that PR is not being worked on.

Thank you! FWIW, those patches and related PRs will all see attention as @karkhaz is bringing them back to life (starting from #4332).

File: src/util/pointer_predicates.h [hunk 2]
```diff
@@ -12,6 +12,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_UTIL_POINTER_PREDICATES_H
 #define CPROVER_UTIL_POINTER_PREDICATES_H
 
+#include "std_expr.h"
 #define SYMEX_DYNAMIC_PREFIX "symex_dynamic::"
 
 class exprt;
```
Comment (tautschnig) [inline src/util/pointer_predicates.h]: Some of those forward declarations will now be unnecessary given the added include.

File: src/util/pointer_predicates.h [hunk 3]
```diff
@@ -33,7 +33,6 @@ exprt good_pointer_def(const exprt &pointer, const namespacet &);
 exprt null_object(const exprt &pointer);
 exprt null_pointer(const exprt &pointer);
 exprt integer_address(const exprt &pointer);
-exprt is_invalid_pointer(const exprt &pointer);
```
Comment (tautschnig) [inline src/util/pointer_predicates.h]: I guess that's ok, though generally we should first mark such functions as deprecated.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a9774ca9693696540cfb5d28a4e5999debe14c2b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106877160

**Pull request merged**